Time stamp for sample using ATOMIC_BLOCK

TimMathias opened the pull request:
Placed the read of the input pin and the read of the millis timer within an ATOMIC_BLOCK to ensure an accurate time stamp for the sample. Also, only read the millis timer once per sample, and use this time stamp for subsequent time difference calculations.

TimMathias committed:
Update DebouncerImpl.h

Call millis() as close as possible to digitalRead() to ensure that the sample time coincides with the sample value. And only call millis() once so that subsequent calculations are based upon the time of the sample. Added a few const keywords.

TimMathias committed:
Update DebouncerImpl.h

Added ATOMIC_BLOCK around critical timing code in update() method of Debouncer class.

TimMathias committed:
Update DebouncerImpl.h

Removed call to nointerrupts() within ATOMIC_BLOCK because it's not needed.

TimMathias committed:
Corrected an erroneous edit.

Changed the erroneous int8_t back to the original uint8_t in the Debouncer constructor.

hideakitai commented:
@TimMathias Thanks for your PR! (and fixing consts...!) Can you use `#ifdef SOME_BOARDS_OR_ARCH` to isolate the boards? Since I think `atomic` is limited to some boards that can be used.

hideakitai commented on Debouncer/DebouncerImpl.h:
```diff
@@ -69,21 +70,29 @@ class Debouncer
 
     void update()
     {
         bool curr_state = digitalRead(pin_target);
```
```suggestion
        bool curr_state = digitalRead(pin_target);
        uint32_t curr_ms = millis();
```
And I also think these lines will be better just to init the variables (no `digitalRead()`).

TimMathias replied:
Good idea.

TimMathias replied:
Regarding `#ifdef SOME_BOARDS_OR_ARCH`, do you know which boards do and do not support `ATOMIC_BLOCK`?

I've also asked here:
https://github.com/arduino/reference-en/issues/761

TimMathias committed:
Update DebouncerImpl.h

TimMathias committed:
Update DebouncerImpl.h

TimMathias committed:
Added conditional compilation for ATOMIC_BLOCK

TimMathias commented:
I've added conditional compilation for the `ATOMIC_BLOCK` macro. It tests to see whether `ATOMIC_BLOCK` is defined, rather than testing for the architecture. Is this acceptable?

hideakitai commented:
@TimMathias Sorry for late reply. That's cool, but we can't include `util/atomic` anyway...
Which board do you use? I couldn't compile that in ESP32.
I will check which board can use atomic later when I have time, but if you can, please let me know which is capable board just as long as you know.

hideakitai commented:
@TimMathias Mmm... I have checked some boards but maybe `util/atomic.h` and  `ATOMIC_BLOCK` is defined only on AVR boards and it is not general for Arduino. So I think you should use `sei()` `cli()` instead of `ATOMIC_BLOCK`. And your PR is not compilable because `input_state` and `input_pin` is not defined...
In addition to that, interrupt routine won't take so much time (more than 1ms) generally, so it is too much to disable interrupts here, and not critical I think.
Instead of that, I think it is better to use `ATOMIC_BLOCK` or `sei()` `cli()` for `Debouncer::update()` call inside your `.ino` file,
How do you think about it?

TimMathias committed:
Update DebouncerImpl.h

Moved the `atomic.h` include into an `__AVR__` conditional compilation block.
Corrected `input_pin` to `pin_target`.
Corrected `current_ms` to `curr_ms`.

TimMathias commented:
I've corrected the undefined variables, and moved the `atomic.h` include into an AVR conditional compilation block.

The advantage of `ATOMIC_BLOCK(ATOMIC_RESTORESTATE)` is that is saves the interrupt state when entering the block and restores it when leaving. It doesn't need to know whether interrupts are enabled or disabled - it just saves and restores. The disadvantage is that it is not defined for all architectures.

The issue with `sei()` and `cli()` if used within a library is that the current interrupt state is unknown<sup>[1]</sup> within the library, so enabling interrupts upon exit may be the wrong thing to do if the caller from the user programme has disabled interrupts.

If `update()` is called from an ISR, interrupts are diabled/re-enabled upon entry and exit of the ISR, so it's not an issue. If `update()` is called from a polling loop, the user has to decide what to do for each call for each button/signal, so it would be good to have a method/function that takes care of it internally for every caller.

1. I think it is possible to work it out, but it's different for each architecture. Need more investigation digging through datasheets for every architecture to possibly implement multiple ATOMIC_BLOCK macros - one for each architecture. A series of:

```c++
#ifdef __AVR__
#include <util/atomic.h>
#elif defined(__ARCH2__)
#define ATOMIC_BLOCK(/* for architecture 2 */)
#elif defined(__ARCH3__)
#define ATOMIC_BLOCK(/* for architecture 3 */)
// etc
#endif
.
.
.
#ifdef ATOMIC_BLOCK
    ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
    {
#endif

        // Do not disturb.

#ifdef ATOMIC_BLOCK
    }
#endif
```

hideakitai commented:
@TimMathias Sorry for late reply, I've got your opinion.
But basically, this library want to check if digital pin has really changed, in a long duration for MCUs (like 20-50msec). So the effect to accuracy of timestamp by disabling ISR is not big. And the library should be general for most of the platforms (not only for AVR), so I want to ensure that this library does NOT disable ISR. It is not general to disable ISRs only for AVRs. So I prefer to disable ISRs outside the `update()` call explicitly.

For these reasons, I will close this PR, but thanks for cool discussion about that, and correction to `consts` and multiple `millis()` call. I will fix them. Thank you!
